src/host/getset.cpp

```diff
@@ -466,6 +477,61 @@ void ApiRoutines::SetConsoleActiveScreenBufferImpl(SCREEN_INFORMATION& newContex
         LockConsole();
         auto Unlock = wil::scope_exit([&] { UnlockConsole(); });
 
+        auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+
+        if (&newContext.GetActiveBuffer() == &gci.GetActiveOutputBuffer())
+        {
+            return;
+        }
+
+        if (auto writer = gci.GetVtWriter())
+        {
+            const auto viewport = gci.GetActiveOutputBuffer().GetBufferSize();
+            const auto size = viewport.Dimensions();
+            const auto area = static_cast<size_t>(viewport.Width() * viewport.Height());
+
+            auto& main = newContext.GetMainBuffer();
+            auto& alt = newContext.GetActiveBuffer();
+            const auto hasAltBuffer = &alt != &main;
+
+            // TODO GH#5094: This could use xterm's XTWINOPS "\e[8;<height>;<width>t" escape sequence here.
+            THROW_IF_NTSTATUS_FAILED(main.ResizeTraditional(size));
+            main.SetViewportSize(&size);
+            if (hasAltBuffer)
+            {
+                THROW_IF_NTSTATUS_FAILED(alt.ResizeTraditional(size));
+                alt.SetViewportSize(&size);
+            }
+
+            Viewport read;
+            til::small_vector<CHAR_INFO, 1024> infos;
+            infos.resize(area, CHAR_INFO{ L' ', FOREGROUND_BLUE | FOREGROUND_GREEN | FOREGROUND_RED });
+
+            const auto dumpScreenInfo = [&](SCREEN_INFORMATION& screenInfo) {
+                THROW_IF_FAILED(ReadConsoleOutputWImpl(screenInfo, infos, viewport, read));

```

Comment:
Yeah - this is going to be the root cause of a lot of our scrolling (and similar) issues. We *have* the ability to convert part of the buffer into a full-fidelity VT stream, so I don't completely understand why we wouldn't use it.

I know, the eventual final ConPTY API (whose spec I merged, but which is still in flux (and I don't know why I merged it)) will force these kinds of narrowing conversions... but until then, why make ourselves repeatedly sad by doing them?

Reply:
If the goal is to break something, I don't think there's any benefit to delaying the inevitable. Why would we want to accrue more people that rely on the soon-to-be-broken functionality after all?

In this case, however, the breakage is temporary, because a minimal console server API must be able to express console buffer switching.

Reply:
I'm very iffy on this. The breakage is temporary, but **not a short period of time**.

It is possible to get 24-bit color and extended attributes into the console buffer, and it is arguably _incorrect_ for APIs which do not use `CHAR_INFO` to damage that data.

The public surface of the console APIs that use `CHAR_INFO` is by-necessity narrowing. That doesn't mean that our internal interfaces should be narrowing as well.

What, per se, is wrong with applications relying on a mix of APIs? They exist, our users use them, and the users are the people we are punishing. They'll file bugs on us, so we are the people we are punishing. We aren't even punishing the developers (and it is questionable whether they should be punished! What did they do wrong? Progressively enhance their application without rewriting it entirely? That's a _good_ thing.)

Reply:
Is this a theoretical or practical concern? Because as of now there are no known applications that rely on surrogate pairs and extended attributes in combination with `SetConsoleActiveScreenBuffer`.

The VT serialization API we added for persistence to `TextBuffer` is no different to having a `VtEngine`, which is something we're trying to move away from due to its fragility. In fact, I'd go as far as to say that the persistence logic is just as fragile as `VtEngine`. If we now rely on that API for these console APIs, we've not improved our stability and predictability _at all_. The easiest examples to demonstrate this is how this won't properly preserve marks or sixels.

The way I see it, preserving extended attributes during `SetConsoleActiveScreenBuffer` is a red herring, because the current implementation has significantly worse flaws that make it entirely unsuitable in WT. Most importantly, it doesn't resize the terminal's screen buffer, which results in the entire output to be scribbled over seemingly randomly by the console application. This results in garbled output and causes contents to be scrolled off potentially destroying even existing contents.

Preserving extended attributes during `ScrollConsoleScreenBuffer` on the other hand is solvable in the near term by using sequences like DECCRA. It's just not part of the initial implementation in this PR.

Reply:
I should mention that one of my original ideas when implementing the VT paging operations was that they could potentially be usable as a replacement for `SetConsoleActiveScreenBuffer` et al. They aren't really equivalent concepts, though, so it probably wouldn't be a very clean mapping, but maybe it's still something we could try in the future.

Reply:
Sorry, I had my wires crossed. I was thinking about the vim scrolling issue, and not the alternate console buffer one. :)

Reply:
> I should mention that one of my original ideas when implementing the VT paging operations was that they could potentially be usable as a replacement for `SetConsoleActiveScreenBuffer` et al. They aren't really equivalent concepts, though, so it probably wouldn't be a very clean mapping, but maybe it's still something we could try in the future.

I'll make sure to note this down in follow-up improvements that we can make!

> Sorry, I had my wires crossed. I was thinking about the vim scrolling issue, and not the alternate console buffer one. :)

(I hope my tone was alright.)

---

I'll try to make it the best VT translation that I can possibly make, so I hope that I can address both scrolling and buffer swapping in the future. Even if this new implementation is worse in some subtle regards, my hope is that it's still overall a significant improvement to what we had. I don't think I can solve them until 1.22, but I think it'll still be fine. I suspect the number of complaints we'll receive will be either appropriate to or less than what we may expect given how much this changes the ConPTY architecture. Long-term I think this is still a significantly better foundation on which we can solve the issues with those 2 APIs.